### ISSUE TITLE
Cannot Serialize Object into State

### ISSUE THREAD

**Issue by seanaye** [oakserver/oak#361]
When if I initialize the Oak application with some state that is a closure then the app crashes on the first request it receives with `error: Uncaught (in promise) TypeError: Invalid argument
  return core ? core.deserialize(core.serialize(value)) : cloneValue(value);`

App setup
```ts
  const authClient = getAuth0Client({
    clientId: Deno.env.get("AUTH0_CLIENT_ID"),
    clientSecret: Deno.env.get("AUTH0_CLIENT_SECRET"),
  });
  const stripeClient = getStripeClient({
    stripeKey: Deno.env.get("STRIPE_KEY"),
  });
  console.log({ authClient, stripeClient });
  const app = new Application({
    state: {
      authClient,
      stripeClient,
    },
  });
```

the source for the `getClient` code is at https://github.com/seanaye/connectors

How else should I give middleware access to "global" api clients

EDIT:
Oak 7.7.0
deno 1.11.5 (release, aarch64-apple-darwin)
v8 9.1.269.35
typescript 4.3.2

**Comment by kitsonk** [comment]
This is a bit of a challenge that I have been thinking about.  We changed the behaviour to "clone" the state for each context, so that people wouldn't stash stuff in there that "leaked" information from one request to the next, assuming that people would limit it to just simple objects that could easily be cloned.  I guess that assumption was incorrect.

I can see the utility of stashing some things in the app state for sharing from middleware that persists for the lifetime of the app, but I can also see the need to have data that is shared only within a specific context, but that you might want to "seed" that data.

Maybe the right thing to do is use `app.state` as the prototype for `ctx.state`, meaning that mutations to the _shallow_ values would only affect the `ctx.state`.  The other option would be that `ctx.state` is a proxy of `app.state` and then there could be finer grained control of what get changes.

Basically I need to think about this for a bit longer.

**Comment by kitsonk** [comment]
I did some thinking about this, and did the following:

- "swallowed" errors on cloning, so the creating a new context shouldn't throw, but non-cloneable objects will get skipped.
- Provided four options to indicate how the context's `.state` should be created which can be set when creating the application using the `contextState` option:
  - `"clone"` this is the new default behaviour, which tries to clone the values in the application's `.state` and skips those that it can't clone. It is effectively the same behaviour that occurs today, but without the halting errors.  This means that a context's `.state` will only endure for the lifetime of context.
  - `"alias"` is the way it used to work, where the application `.state` and the context's `.state` are the same object.
  - `"prototype"` create a new object for the context's state with the applications `.state` as the prototype. For those comfortable with how prototypes work in JavaScript, this will make sense, and allows "default" properties to be set, but mutations to the properties of the context's `.state` to only endure for the lifetime of the context, but _deeper_ modifications to non-primatives affect the applications `.state`.
  - `"empty"` an empty `.state` is created for each context and only lives for the lifetime of the context.

It should be noted that the context has `.app` which is a reference to the application, so `.app.state` is available for things that should endure for the lifetime of the application and can be accessed by middleware.

**Comment by seanaye** [comment]
@kitsonk thanks for your great work supporting this library.
